url9ecf1e8297_nhamidi_ninja_tgz-pJ8-net_fec_openrq_test_TestRunnerJ8/src/RQLibrary/Encoder.java

```diff
@@ -0,0 +1,2552 @@
+		
+		intermediate_symbols = generateIntermediateSymbols(sb);
+		
+	    } catch (SingularMatrixException e) {
+		// this never happens here
+		System.exit(-42437);
+	    }
+	    
+	    /*
+	     * Second encoding step
+	     */
+	    
+	    // sending original source symbols
+	    int source_symbol;
+	    int source_symbol_index;
+	    
+	    for (source_symbol = 0, source_symbol_index = 0; source_symbol < K; source_symbol++, source_symbol_index += sb.getT()) {
+		encoded_symbols[source_symbol] = new EncodingSymbol(SBN, source_symbol, Arrays.copyOfRange(ssymbols, source_symbol_index,
+			(int) (source_symbol_index + sb.getT())));
+	    }
+	    
+	    // generating/sending repair symbols
+	    for (int repair_symbol = 0; repair_symbol < numRepairSymbols; repair_symbol++) {
+		int isi = kLinha + repair_symbol;
+		int esi = K + repair_symbol;
+		
+		byte[] enc_data = enc(kLinha, intermediate_symbols, new Tuple(kLinha, isi));
+		
+		encoded_symbols[source_symbol + repair_symbol] = new EncodingSymbol(SBN, esi, enc_data);
+	    }
+	    
+	    encoded_blocks[source_block_index] = new EncodingPacket(SBN, encoded_symbols, K, sb.getT());
+	}
+	
+	return encoded_blocks;
+    }
+    
+    /**
+     * Initializes the G_LDPC1 submatrix.
+     * 
+     * @param constraint_matrix
+     * @param B
+     * @param S
+     */
+    private static void initializeG_LPDC1(byte[][] constraint_matrix, int B, int S) {
+	int circulant_matrix = -1;
+	
+	for (int col = 0; col < B; col++) {
+	    int circulant_matrix_column = col % S;
+	    
+	    if ( circulant_matrix_column != 0 ) {
+		// cyclic down-shift
+		constraint_matrix[0][col] = constraint_matrix[S - 1][col - 1];
+		
+		for (int row = 1; row < S; row++) {
+		    constraint_matrix[row][col] = constraint_matrix[row - 1][col - 1];
+		}
+	    } else { // if 0, then its the first column of the current circulant matrix
+	    
+		circulant_matrix++;
+		
+		// 0
+		constraint_matrix[0][col] = 1;
+		
+		// (i + 1) mod S
+		constraint_matrix[(circulant_matrix + 1) % S][col] = 1;
+		
+		// (2 * (i + 1)) mod S
+		constraint_matrix[(2 * (circulant_matrix + 1)) % S][col] = 1;
+	    }
+	}
+    }
+    
+    /**
+     * Initializes the G_LPDC2 submatrix.
+     * 
+     * @param constraint_matrix
+     * @param S
+     * @param P
+     * @param W
+     */
+    private static void initializeG_LPDC2(byte[][] constraint_matrix, int S, int P, int W) {
+	
+	for (int row = 0; row < S; row++) {
+	    // consecutives 1's modulo P
+	    constraint_matrix[row][(row % P) + W] = 1;
+	    constraint_matrix[row][((row + 1) % P) + W] = 1;
+	}
+    }
+    
+    /**
+     * Initializes the I_S submatrix.
+     * 
+     * @param constraint_matrix
+     * @param S
+     * @param B
+     */
+    private static void initializeIs(byte[][] constraint_matrix, int S, int B) {
+	
+	for (int row = 0; row < S; row++) {
+	    for (int col = 0; col < S; col++) {
+		if ( col != row )
+		    continue;
+		else
+		    constraint_matrix[row][col + B] = 1;
+	    }
+	}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 977:984, 999:1006.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.